docs/src/ref/distributions.md

```diff
@@ -19,3 +17,207 @@ exponential
+   and `x_i` are all of type `CONST`, `f(x_1, x_2, ...)` is of type `CONST`.
+
+5. **Functions of `CONST`s and `ARG`s are `ARG`s.** If `f` is a
+   _differentiable_ function, and each `x_i` is either a `CONST` or a _scalar_
+   `ARG` (with at least one `x_i` being an `ARG`), then `f(x_1, x_2, ...)` is
+   of type `ARG`.
+
+6. **Functions of `CONST`s, `ARG`s, and `RND`s are `RND`s.** If `f` is one of a
+   special set of deterministic functions we've defined (`+`, `-`, `*`, `/`,
+   `exp`, `log`, `getindex`), and exactly one of its arguments `x_i` is of type
+   `RND`, then `f(x_1, x_2, ...)` is of type `RND`.
+
+One way to think about this, without all the rules, is that `CONST` values are
+"contaminated" by interaction with `ARG` values (becoming `ARG`s themselves),
+and both `CONST` and `ARG` are "contaminated" by interaction with `RND`.
+Thinking of the body as an AST, the journey from leaf node to root node always
+involves transitions in the direction of `CONST -> ARG -> RND`, never in
+reverse. (There are certain similarities here with [security
+types](https://en.wikipedia.org/wiki/Security_type_system), which also enforce
+a single direction of information flow -- anything that touches classified data
+is also classified. Also, Tabular v2's type system, which similarly separates
+deterministic and random values in its type system.)
+
+#### Restrictions
+Users may _not_ reassign to arguments (like `x` in the above example), and may
+not apply functions with side effects. Names bound to expressions of type `RND`
+must be used only once. e.g., `let x = normal(0, 1) in x + x` is not allowed.
+
+**TODO:** (Expand on scalar-vs.-vector problems here.)
+
+#### Examples
+
+Let's walk through some examples.
+
+```julia
+@dist f(x) = exp(normal(x, 1))
+```
+
+We can annotate with types:
+```
+1 :: CONST		  (by rule 1)
+x :: ARG 		  (by rule 2)
+normal(x, 1) :: RND 	  (by rule 3)
+exp(normal(x, 1)) :: RND  (by rule 6)
+```
+
+Here's another:
+```
+@dist function labeled_cat(labels, probs)
+	index = categorical(probs)
+	labels[index]
+end
+```
+
+And the types:
+```
+probs :: ARG 			(by rule 2)
+categorical(probs) :: RND 	(by rule 3)
+index :: RND 			(Julia assignment)
+labels :: ARG 			(by rule 2)
+labels[index] :: RND 		(by rule 6, f == getindex)
+```
+
+Note that `getindex` is designed to work on anything indexible, not just
+vectors. So, for example, it also works with Dicts.
+
+Another one (not as realistic, but it uses all the rules):
+
+```
+@dist function weird(x)
+  log(normal(exp(x), exp(x))) + (x * (2 + 3))
+end
+```
+
+And the types:
+
+```
+2, 3 :: CONST 						(by rule 1)
+2 + 3 :: CONST 						(by rule 4)
+x :: ARG 						(by rule 2)
+x * (2 + 3) :: ARG 					(by rule 5)
+exp(x) :: ARG 						(by rule 5)
+normal(exp(x), exp(x)) :: RND 				(by rule 3)
+log(normal(exp(x), exp(x))) :: RND 			(by rule 6)
+log(normal(exp(x), exp(x))) + (x * (2 + 3)) :: RND 	(by rule 6)
+```
+
+## A note for the long-term future
```

Comment:
What do you see as the benefit for putting this in the Gen documentation?

Reply:
The benefit was that users (who may be researchers in this area) can be aware of the discussion and participate if they want.  But as you pointed out above, an issue is a better place for this.  Filed https://github.com/probcomp/Gen/issues/201.